# Импортитуем библиотеки

In [ ]:
import pandas as pd
import random as rnd
import numpy as np

#Для генерации данных в random необходимо установить seed = "АДВИ_2023", всего длинна каждого списка N = 1000

In [ ]:
rnd.seed('АДВИ_2023') # Фиксируем рандомизацю

#index колонка: создать список из N последовательных целых элементов от 1 до 1000 включительно, далее, с помощью random, перемешать его -> будем использовать как

In [ ]:
index = [i for i in range(1, 1001)] # Создаем списк index 
rnd.shuffle(index) # Перемешиваем список index

In [ ]:
print(index[:5]) # Проверяем первые 5 значений (по итогам просмотра чата должно получиться [476,315,490,748,471])
print(len(index)) # Проверяем длину списка index. Должно быть 1000 элементов.

[476, 315, 490, 748, 471]
1000


#groups колонка: создать список из N элементов, состоящий из значений из списка ["группа_1", "группа_2", "группа_3", "группа_4", "группа_5"] (равномерное распределение);

In [ ]:
population = ['группа_1', 'группа_2', 'группа_3', 'группа_4', 'группа_5'] # Создаем исходный список групп
groups = rnd.choices(population=population, k=1000) # Генерируем список с помощью метода выбора k-массива из исходного списка
groups[:5] # Проверяем, должно получиться [1, 5, 4, 3, 5]

['группа_1', 'группа_5', 'группа_4', 'группа_3', 'группа_5']

# uniform колонка: создать список из N элементов, состоящий из целых значений из равномерного распределения [50, 100]

In [ ]:
uniform = [rnd.randint(50, 100) for i in range(1, 1001)] # Генерим список из 1000 рандомных ЦЕЛЫХ чисел в диапазоне 50<=N<=100
uniform[:5] # Проверим первые 5 чисел. Должно получиться [95, 77, 96, 56, 52]

[95, 77, 96, 56, 52]

# gauss_1, gauss_2, gauss_3 колонки: Создать 3 списка из N элементов каждый, заполненные значениями из нормальных распределений с параметрами: mu = 0, sigma^2 = 1; mu = 0, sigma^2 = 144, mu = 50, sigma^2 = 81

In [ ]:
def gauss_list(mu, sigma): # Создадим функцию по генерированию рандомного списка (чтобы не хардкодить)
  return [rnd.gauss(mu, sigma) for i in range(1, 1001)]

In [ ]:
gauss_1 = gauss_list(0, 1) # Сгенерим рандомный список из 1000 элементов из Гауссовского распределения с параметрами mu = 0, sigma = 1 (в модуле random.gauss sigma без квадрата)
gauss_1[:5] # Проверим первые 5 индексов, должно получиться [3.01029133931237, 0.19027733260392263, -0.2821614316869885, 1.127080293462533, -1.238270197990803]

[3.01029133931237,
 0.19027733260392263,
 -0.2821614316869885,
 1.127080293462533,
 -1.238270197990803]

In [ ]:
gauss_2 = gauss_list(0, 12)
gauss_2[:5]

[2.156346051014784,
 -9.9345500018044,
 9.21107054467491,
 -12.391792676825144,
 10.938748375443827]

In [ ]:
gauss_3 = gauss_list(50, 9)
gauss_3[:5]

[57.45585127992099,
 62.69417653984917,
 54.68889491891202,
 30.1923532867853,
 52.68205247669375]

# Создать pandas DataFrame, где (1) будет индексом, (2)-(4) колонки, названия колонок ["groups", "uniform", "gauss_1", "gauss_2", "gauss_3"] соответственно, не забывая проверить, что типы колонок должны соответствовать заявленным типам данных

In [ ]:
df = pd.DataFrame(index = index, data={'groups':groups, 'uniform':uniform, 'gauss_1':gauss_1, 'gauss_2':gauss_2, 'gauss_3':gauss_3}) # Создадим датафрейм с ранее сгенерированными колонками. Индексами будет выступать колонка index
df.head(10) # Выведем первые 10 строк

groups  uniform   gauss_1    gauss_2    gauss_3
476  группа_1       95  3.010291   2.156346  57.455851
315  группа_5       77  0.190277  -9.934550  62.694177
490  группа_4       96 -0.282161   9.211071  54.688895
748  группа_3       56  1.127080 -12.391793  30.192353
471  группа_5       52 -1.238270  10.938748  52.682052
907  группа_2       70 -1.227963  12.721383  55.538879
457  группа_2       70 -0.553928  -1.674862  48.011108
826  группа_1       73  0.581783  -2.170896  49.992722
5    группа_3       64  1.612765  -1.242515  41.876322
646  группа_3       86 -0.485014  10.381790  55.025045

In [ ]:
df.dtypes # Проверим типы данных

groups      object
uniform      int64
gauss_1    float64
gauss_2    float64
gauss_3    float64
dtype: object

#Для колонок gauss_1 и gauss_2 заменить значения на None: Для gauss_1, если индекс у значение делится на 121 без остатка; Для gauss_2, если дробная часть больше 0.95

In [ ]:
df.loc[df.index % 121 == 0, 'gauss_1'] = None # Заменим значения в gauss_1 по условию
df['gauss_1'].value_counts() # Проверим, появились ли значения None и сколько их

 3.010291    1
 2.033302    1
 0.559920    1
-0.691854    1
 1.180933    1
            ..
-0.394211    1
-1.855531    1
 0.445405    1
 1.253340    1
 0.093496    1
Name: gauss_1, Length: 992, dtype: int64

In [ ]:
df.loc[(df['gauss_2'] > 0) & (df['gauss_2'] % 1 > 0.95), 'gauss_2'] = None # Заменим значения в gauss_2 по условию
df['gauss_2'].value_counts() # Проверим значения в столбце. Если их меньше 1000, значит какие-то значения None

 2.156346     1
 10.827245    1
 8.007369     1
-3.303095     1
 2.818864     1
             ..
-6.189221     1
-29.290553    1
-21.874444    1
 8.921171     1
-9.164475     1
Name: gauss_2, Length: 972, dtype: int64

# Теперь, для пропущенных значений gauss_1 и gauss_2: Для gauss_2 None заменить на среднее по колонке; Удалить строчки из таблицы, где gauss_1 == None.

In [ ]:
df['gauss_2'].fillna(df['gauss_2'].mean(), inplace=True) # Заменим None значения средним по столбцу
df['gauss_2'].value_counts() # Проверим значения

-0.434114     28
 2.156346      1
 0.615091      1
-3.303095      1
 2.818864      1
              ..
-6.189221      1
-29.290553     1
-21.874444     1
 8.921171      1
-9.164475      1
Name: gauss_2, Length: 973, dtype: int64

In [ ]:
df.dropna(subset = ['gauss_1'], inplace=True) # Удалим строки из таблицы по условию, что значение в столбце gauss_1 = None
df.info() # Проверим мощность датасета

<class 'pandas.core.frame.DataFrame'>
Int64Index: 992 entries, 476 to 587
Data columns (total 5 columns):
 #   Column   Non-Null Count  Dtype  
---  ------   --------------  -----  
 0   groups   992 non-null    object 
 1   uniform  992 non-null    int64  
 2   gauss_1  992 non-null    float64
 3   gauss_2  992 non-null    float64
 4   gauss_3  992 non-null    float64
dtypes: float64(3), int64(1), object(1)
memory usage: 46.5+ KB


# Вопросы из теста

Сколько всего получилось записей в таблице после выполнения всех пунктов из условия?

In [ ]:
df.shape # строки х столбцы

(992, 5)

Для колонки gauss_2 посчитать среднее и среднеквадратическое отклонение.

Результаты необходимо написать в одну ячейку, разделенные пробелом, сначала среднее потом среднеквадратическое отклонение. Результаты округлить до сотых. Для дробной части использовать точку, как разделитель.


In [ ]:
df['gauss_2'].mean().round(2) # Среднее

-0.47

In [ ]:
df['gauss_2'].std().round(2) # Среднеквадратическое отклонение

12.07

Какая группа встречается чаще всего и сколько раз?

Сначала вывести название группы, через пробел указать целое число.

In [ ]:
df['groups'].value_counts().sort_values(ascending=False)

группа_4    222
группа_1    215
группа_2    194
группа_3    184
группа_5    177
Name: groups, dtype: int64

Для группа_5 вывести самый "ранний" (наименьший) индекс.

In [ ]:
df['groups'][df['groups'] == 'группа_5'].index.sort_values()

Int64Index([   6,    7,   18,   21,   22,   28,   38,   44,   46,   52,
            ...
             949,  954,  960,  969,  970,  971,  978,  989,  997, 1000],
           dtype='int64', length=177)

Сколько записей из колонки uniform имеют значение не ниже значения 90% перцентиля?

In [ ]:
df[df['uniform'] >= df['uniform'].quantile(0.9)].count()

groups     118
uniform    118
gauss_1    118
gauss_2    118
gauss_3    118
dtype: int64

У какой группы наибольшее значение медианы для gauss_3?

Вывести это значение вместе с группой. Сначала группа, через пробел значение медианы (результат округлить до сотых , для дробной части использовать точку, как разделитель)

In [ ]:
df.groupby('groups')['gauss_3'].median().sort_values(ascending = False).round(2)

groups
группа_2    49.99
группа_5    49.94
группа_1    49.88
группа_3    49.46
группа_4    49.07
Name: gauss_3, dtype: float64

У какой группы наибольшее минимальное значение uniform?

Если такого нет, выбрать группу наибольшим минимальным значением в gauss_1.

Ответ дать в виде группы, через пробел указать значение в gauss_2 с аналогичной агрегацией (результат округлить до сотых, для дробной части использовать точку, как разделитель)

In [ ]:
df.groupby('groups')['uniform'].min()

groups
группа_1    50
группа_2    50
группа_3    50
группа_4    50
группа_5    50
Name: uniform, dtype: int64

In [ ]:
min_value = df.groupby('groups')['gauss_1'].min()
min_value = min_value.max()
min_value

-2.6067727433396337

In [ ]:
#df.groupby('groups')['gauss_1'].min().round(2).sort_values()

In [ ]:
df.loc[df['gauss_1'] == min_value]

groups  uniform   gauss_1   gauss_2    gauss_3     gauss
939  группа_3       62 -2.606773  9.300478  36.542984  0.183173

In [ ]:
df[df['groups'] == 'группа_3']['gauss_2'].min().round(2)

-36.19

Получается, что ответ группа_3 -36.19

Создать новую колонку gauss где значения будут получаться из (gauss_1 + gauss_2) / gauss_3

Чему будет равен 70% перцентиль (результат округлить до сотых, для дробной части использовать точку, как разделитель)?

In [ ]:
df['gauss'] = (df['gauss_1'] + df['gauss_2']) / df['gauss_3']
df['gauss'].quantile(0.7).round(2)

0.11